openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py

```diff
@@ -0,0 +1,210 @@
+"""
+ObjectTags for System-defined Taxonomies
+"""
+from enum import Enum
+from typing import List
+
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.core.exceptions import FieldDoesNotExist
+from django.db import models
+
+
+from openedx_tagging.core.tagging.models import (
+    Taxonomy,
+    OpenObjectTag,
+    ClosedObjectTag,
+)
+from openedx_tagging.core.tagging.registry import register_object_tag_class
+
+
+class SystemDefinedIds(Enum):
+    """
+    System-defined taxonomy IDs
+    """
+    LanguageTaxonomy = 1
+
+
+class SystemDefinedObjectTagMixin:
+    """
+    Mixing for ObjectTags used on all system defined taxonomies
+
+    `system_defined_taxonomy_id``is used to connect the 
+    ObjectTag with the system defined taxonomy.
+    This is because there can be several ObjectTags
+    for the same Taxonomy, ex:
+    
+    - LanguageCourseObjectTag
+    - LanguageBlockObjectTag
+
+    On the example, there are ObjectTags for the same Language
+    taxonomy but with different objects.
+
+    Using this approach makes the connection between the ObjectTag
+    and system defined taxonomy as hardcoded and can't be changed.
+    """
+
+    system_defined_taxonomy_id = None
+
+    def _check_system_taxonomy(self, taxonomy: Taxonomy = None):
+        """
+        Validates if the taxonomy is system-defined and match
+        with the name stored in the object tag
+        """
+        return (
+            bool(taxonomy) and
+            taxonomy.system_defined and
+            taxonomy.id == self.system_defined_taxonomy_id
+        )
+
+
+class OpenSystemObjectTag(OpenObjectTag, SystemDefinedObjectTagMixin):
+    """
+    Free-text object tag used on system-defined taxonomies
+    """
+
+    class Meta:
+        proxy = True
+
+    def _check_taxonomy(self):
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
+
+    
+class ClosedSystemObjectTag(ClosedObjectTag, SystemDefinedObjectTagMixin):
+    """
+    Object tags linked to a closed system-taxonomy
+    """
+
+    class Meta:
+        proxy = True
+
+    def _check_taxonomy(self):
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
+
+
+class ModelObjectTag(OpenSystemObjectTag):
```

Comment:
@ChrisChV Since we've gone back to subclassing Taxonomy again, I wonder if we could create a ModelTaxonomy class, which can:

* Create Tag objects on the fly when new ObjectTags are added
   * set external_id=model.pk
   * set value = value we want to show to the world, e.g. User.username or Organization.name
   * "Resync" could somehow re-fetch the related object and update this value? Or maybe we need another method.
* Can the ModelTaxonomy can be "closed" even though the Tag values don't exist (yet)?
* Can we return "autocomplete_tags" based on the values available in the linked model?

Reply:
> Create Tag objects on the fly when new ObjectTags are added

This has been one of the [rejected options](https://github.com/openedx/openedx-learning/blob/main/docs/decisions/0012-system-taxonomy-creation.rst#tags-created-by-auto-generated-from-the-codebase) for creating Tags. But now that I see it, it's easier than it seemed and has many more advantages. Yes, I will build this type of taxonomy

Reply:
@pomegranited I updated the ADR here https://github.com/open-craft/openedx-learning/pull/3/commits/4e675bf50fa9756b9bf65c9be9c40f4b4b4b8fc5